packages/langium/test/grammar/type-system/inferred-types.test.ts

```diff
@@ -543,6 +543,44 @@ describe('types of `$container` and `$type` are correct', () => {
             }
         `);
     });
+
+    test('types of `$type` for declared deep hierarchies', async () => {
```

Comment:
Suggestion: It might make sense to split this test into multiple tests that test for the following scenarios:

1. Linear (`C extends B, B extends A`)
2. Deviating (`B extends A, C extends A`)
3. Joining (`C extends A and B`)
4. A mix of all of the above.